buildSrc/src/main/java/org/elasticsearch/gradle/test/rest/CopyRestApiTask.java

```diff
@@ -0,0 +1,198 @@
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFiles;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.SkipWhenEmpty;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.TaskAction;
+import org.gradle.api.tasks.util.PatternFilterable;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.internal.Factory;
+
+import javax.inject.Inject;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Copies the files needed for the Rest YAML specs to the current projects test resources output directory.
+ * This is intended to be be used from {@link CopyRestApiPlugin} since the plugin wires up the needed
+ * configurations and custom extensions.
+ * @see CopyRestApiPlugin
+ */
+public class CopyRestApiTask extends DefaultTask {
+    private static final Logger logger = Logging.getLogger(CopyRestApiTask.class);
+    private static final String COPY_TO = "rest-api-spec/api";
+    final ListProperty<String> includeCore = getProject().getObjects().listProperty(String.class);
+    final ListProperty<String> includeXpack = getProject().getObjects().listProperty(String.class);
+
+    Configuration coreConfig;
+    Configuration xpackConfig;
+
+    private final PatternFilterable corePatternSet;
+    private final PatternFilterable xpackPatternSet;
+
+    public CopyRestApiTask() {
+        corePatternSet = getPatternSetFactory().create();
+        xpackPatternSet = getPatternSetFactory().create();
+    }
+
+    @Inject
+    protected Factory<PatternSet> getPatternSetFactory() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Input
+    public ListProperty<String> getIncludeCore() {
+        return includeCore;
+    }
+
+    @Input
+    public ListProperty<String> getIncludeXpack() {
+        return includeXpack;
+    }
+
+    @SkipWhenEmpty
+    @InputFiles
+    public FileTree getInputDir() {
+        xpackPatternSet.setIncludes(includeXpack.get().stream().map(prefix -> prefix + "*/**").collect(Collectors.toList()));
+        ConfigurableFileCollection fileCollection = getProject().files(xpackConfig.getAsFileTree().matching(xpackPatternSet));
+        if (BuildParams.isInternal()) {
+            corePatternSet.setIncludes(includeCore.get().stream().map(prefix -> prefix + "*/**").collect(Collectors.toList()));
+            fileCollection.plus(coreConfig.getAsFileTree().matching(corePatternSet));
+        } else {
+            fileCollection.plus(coreConfig);
+        }
+        // if project has rest tests or the includes are explicitly configured execute the task, else NO-SOURCE due to the null input
+        return projectHasYamlRestTests() || includeCore.get().isEmpty() == false || includeXpack.get().isEmpty() == false
+            ? fileCollection.getAsFileTree()
+            : null;
+    }
+
+    @OutputDirectory
+    public File getOutputDir() {
```

Comment:
FYI, we now have three (3) tasks all using the same output directory which is going to cause issues with incremental build. I'll address this in a subsequent PR since it's a bit involved to register additional sourceset output directories but it's something to be aware of, by convention multiple tasks dumping their output in the same place is bad practice.

Reply:
Which 3 are you referring to ? 

I think this is the only task that is writing to `<outputResourcDir>/rest-api-spec/api`, the other new task writes to `<outputResourcDir>/rest-api-spec/test` and the old tasks have been removed. Inside this task, there are 3 `project.copy` , is that what you mean ?

Reply:
Primarily the `processTestResources` task. Since it's output is the root resources directory, both these tasks put stuff in there, which is going to confusing up-to-date checking for that task.

Reply:
Ah, I see thanks for the explanation and the follow up PR (please CC me on that change).